From fb9142925cfbe27b70f181ce70e68df9b13f3152 Mon Sep 17 00:00:00 2001
From: Weng Xuetian <wengxt@gmail.com>
Date: Sat, 11 May 2024 13:34:55 -0700
Subject: [PATCH] Fix icon name mapping in flatpak (#1049)

Now we have a much of icons that uses "fcitx_" as the name, including
mozc, skk, kkc, rime, to avoid xdg icon spec based fallback.

But we missed the flatpak naming mapping for those icons. Update the
condition here, and also handle the special name "fcitx" (I don't think
we do have code that uses "fcitx" as icon, but well.. )

https://github.com/fcitx/mozc/issues/53#issuecomment-2105739015
---
 src/lib/fcitx/icontheme.cpp | 8 +++++++-
 test/testicontheme.cpp      | 3 +++
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/src/lib/fcitx/icontheme.cpp b/src/lib/fcitx/icontheme.cpp
index a1471beb0..46979bf20 100644
--- a/src/lib/fcitx/icontheme.cpp
+++ b/src/lib/fcitx/icontheme.cpp
@@ -797,7 +797,13 @@ std::string IconTheme::defaultIconThemeName() {
 
 /// Rename fcitx-* icon to org.fcitx.Fcitx5.fcitx-* if in flatpak
 std::string IconTheme::iconName(const std::string &icon, bool inFlatpak) {
-    if (inFlatpak && stringutils::startsWith(icon, "fcitx-")) {
+    constexpr std::string_view fcitxIconPrefix = "fcitx";
+    if (inFlatpak && stringutils::startsWith(icon, fcitxIconPrefix)) {
+        // Map "fcitx" to org.fcitx.Fcitx5
+        // And map fcitx* to org.fcitx.Fcitx5.fcitx*
+        if (icon.size() == fcitxIconPrefix.size()) {
+            return "org.fcitx.Fcitx5";
+        }
         return stringutils::concat("org.fcitx.Fcitx5.", icon);
     }
     return icon;
diff --git a/test/testicontheme.cpp b/test/testicontheme.cpp
index e6935c59f..d18657296 100644
--- a/test/testicontheme.cpp
+++ b/test/testicontheme.cpp
@@ -21,6 +21,9 @@ int main() {
     FCITX_ASSERT(IconTheme::iconName("fcitx-pinyin", false) == "fcitx-pinyin");
     FCITX_ASSERT(IconTheme::iconName("fcitx-pinyin", true) ==
                  "org.fcitx.Fcitx5.fcitx-pinyin");
+    FCITX_ASSERT(IconTheme::iconName("fcitx_mozc", true) ==
+                 "org.fcitx.Fcitx5.fcitx_mozc");
+    FCITX_ASSERT(IconTheme::iconName("fcitx", true) == "org.fcitx.Fcitx5");
 
     for (const auto &inheritTheme : theme.inherits()) {
         FCITX_INFO() << inheritTheme.name().match();
